### PR TITLE
DAOS-3752 container: fetch snapshot list for aggregation

### PR DESCRIPTION
**Pull request by NiuYawei**
Fetch snapshot list in aggregation ULT instead on container target
open, otherwise, aggregation won't start until container is opened.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/iosrv/server_iv.c [hunk 1]
```diff
@@ -919,6 +919,8 @@ iv_op(struct ds_iv_ns *ns, struct ds_iv_key *key, d_sg_list_t *value,
 {
 	int rc;
 
+	if (ns->iv_master_rank == ((d_rank_t)-1))
+		return -DER_NOTLEADER;
```
Comment (wangdi1) [inline src/iosrv/server_iv.c:923]: this seems has been checked inside ivc_on_hash(), which should be called in iv_op_internal()? I wonder if you return here, then for those retry == true case will not retry anymore in this case.

File: src/pool/srv_target.c [hunk 2]
```diff
@@ -278,15 +266,24 @@ pool_alloc_ref(void *key, unsigned int ksize, void *varg,
...
+	collective_arg.pla_map_version = arg->pca_map_version;
+	rc = dss_thread_collective(pool_child_add_one, &collective_arg, 0,
+				   DSS_ULT_IO);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to add ES pool caches: "DF_RC"\n",
+			DP_UUID(key), DP_RC(rc));
+		goto err_iv_ns;
+	}
+
 	*link = &pool->sp_entry;
```
Comment (wangdi1) [inline src/pool/srv_target.c:280]: why move this part? for fixing bug?

File: src/iosrv/server_iv.c (hunk 1 above)
Comment (NiuYawei) [reply src/iosrv/server_iv.c:923]: Have we ever tested the retry logic? If this -1 master rank passed in ivc_on_hash(), it'll crash in iv code.

File: src/pool/srv_target.c (hunk 2 above)
Comment (NiuYawei) [reply src/pool/srv_target.c:280]: I think we should create IV namespace first, then create ds_pool_child which will start all containers and start aggregation ULT (where fetching snapshot list by IV).

File: src/iosrv/server_iv.c (hunk 1 above)
Comment (NiuYawei) [reply src/iosrv/server_iv.c:923]: seems that's a defect which has been fixed in #PR 2700 (added error checking in ds_iv_done()), so probably I can remove this check here after than patch landed.

File: src/container/srv_target.c [hunk 3]
```diff
@@ -78,6 +78,10 @@ cont_aggregate_runnable(struct ds_cont_child *cont)
 	/* snapshot list isn't fetched yet */
 	if (cont->sc_aggregation_max == 0) {
 		D_DEBUG(DB_EPC, "No aggregation before snapshots fetched\n");
+		/* fetch snapshot list */
+		if (dss_get_module_info()->dmi_tgt_id == 0)
+			ds_cont_tgt_snapshots_refresh(cont->sc_pool->spc_uuid,
+						      cont->sc_uuid);
```
Comment (liw) [inline src/container/srv_target.c:84]: Can there be more than one cont_snapshots_refresh_ult ULTs running at the same time? We're creating anonymous ULTs again (but not a fault of this PR).

**Pull request merged**